### PR TITLE
Audit, comment and fix gRPC status code handling.

### PR DESCRIPTION
**Pull request by qingling128**

### CODE REVIEW

File: lib/fluent/plugin/out_google_cloud.rb [hunk 1]
```diff
@@ -602,26 +602,51 @@ def write(chunk)
...
             case error.code
-            when GRPC::Core::StatusCodes::CANCELLED,
-                 GRPC::Core::StatusCodes::UNAVAILABLE,
-                 GRPC::Core::StatusCodes::DEADLINE_EXCEEDED,
-                 GRPC::Core::StatusCodes::INTERNAL,
-                 GRPC::Core::StatusCodes::UNKNOWN
-              # Server error, so retry via re-raising the error.
+            # Server error, so retry via re-raising the error.
+            when \
+                # Maps to http status 500 (Internal Server Error)
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: Nit: s/`http status`/`HTTP status`/ everywhere below.

Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: [Optional] The "Maps to" part is repetitive and seems redundant.

Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: Nit: missing trailing '.'.

File: lib/fluent/plugin/out_google_cloud.rb [hunk 2]
```diff
@@ -602,26 +602,51 @@ def write(chunk)
...
+
+            # Most client errors indicate a problem with the request itself and
+            # should not be retried.
             when \
-                # Most client errors indicate a problem with the request itself
-                # and should not be retried.
-                GRPC::Core::StatusCodes::UNIMPLEMENTED,
+                # Maps to http status 400 (Bad Request).
+                GRPC::Core::StatusCodes::INVALID_ARGUMENT,
+                # Authorization error. These are usually solved via a `gcloud
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: Let's reorder these:
```ruby
                # Authorization error. Maps to HTTP status 401 (Unauthorized).
                # These are usually solved via a `gcloud auth` call, or by
                # modifying the permissions on the Google Cloud project.
```

File: lib/fluent/plugin/out_google_cloud.rb [hunk 3]
```diff
@@ -602,26 +602,51 @@ def write(chunk)
...
+                GRPC::Core::StatusCodes::FAILED_PRECONDITION,
+                # Maps to http status 429 (Too Many Requests).
                 GRPC::Core::StatusCodes::RESOURCE_EXHAUSTED,
-                # Authorization error.
-                # These are usually solved via a `gcloud auth` call, or by
-                # modifying the permissions on the Google Cloud project.
-                GRPC::Core::StatusCodes::UNAUTHENTICATED
+                # Maps to http status 499 (Client Closed Request).
+                GRPC::Core::StatusCodes::CANCELLED,
+                # Maps to the remaining http codes in both 4XX and 5XX category.
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: Nit: `4xx and 5xx` (lowercase "x"). Also below.

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 4]
```diff
@@ -28,9 +28,17 @@ def test_configure_use_grpc
 
   def test_client_error
     setup_gce_metadata_stubs
-    { 8 => 'ResourceExhausted',
-      12 => 'Unimplemented',
-      16 => 'Unauthenticated' }.each_with_index do |(code, message), index|
+    {
+      1 => 'Cancelled',
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: Can this be:
```ruby
    include GRPC::Core::StatusCodes
    ...
    {
      CANCELLED => 'Cancelled',
      UNKNOWN => 'Unknown',
      ...etc
    }
```
?

File: lib/fluent/plugin/out_google_cloud.rb (hunk 1 above)
Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: Changed.

Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: Removed.

Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: Added.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 2 above)
Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: Changed to:

```
                # HTTP status 401 (Unauthorized).
                # These are usually solved via a `gcloud auth` call, or by
                # modifying the permissions on the Google Cloud project.
```

The `Authorization error.` part seems redundant.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 3 above)
Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: Changed.

File: test/plugin/test_out_google_cloud_grpc.rb (hunk 4 above)
Comment (qingling128) [reply test/plugin/test_out_google_cloud_grpc.rb]: Changed.

File: lib/fluent/plugin/out_google_cloud.rb [hunk 5]
```diff
@@ -602,26 +602,50 @@ def write(chunk)
 
           rescue GRPC::BadStatus => error
             increment_failed_requests_count(error.code)
+            # See the mapping between HTTP status and grpc status code at:
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: Nit: `gRPC`.

Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: Done.

File: lib/fluent/plugin/out_google_cloud.rb [hunk 6]
```diff
@@ -666,23 +690,23 @@ def write(chunk)
             end
 
           rescue Google::Apis::ServerError => error
-            # Server error, so retry via re-raising the error.
+            # 5xx Server errors, so retry via re-raising the error.
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: For consistency: `5xx server errors. Retry via re-raising the error.`.

Comment (qingling128) [reply lib/fluent/plugin/out_google_cloud.rb]: Done.

**Comment by qingling128** [comment]
This is released in fluent-plugin-google-cloud [v0.6.11](https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/releases/tag/v0.6.11) and google-fluentd [v1.5.24](https://github.com/GoogleCloudPlatform/google-fluentd/releases/tag/v1.5.24).

**Pull request merged**